### PR TITLE
feat(plugin-meetings): rejoin after inactive

### PR DESCRIPTION
**Pull request by arun3528**
Screen Shot 2020-06-16 at 1 44 21 PM

# Pull Request Template
Rejoin the existing meeting after meeting inactive comes in. 

## Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Test Coverage

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

**Test Configuration**:
* SDK Version
* Node/Browser Version
* NPM Version

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: packages/node_modules/@webex/plugin-meetings/src/reconnection-manager/index.js [hunk 1]
```diff
@@ -180,11 +194,16 @@ export default class ReconnectionManager {
   media(meeting) {
     this.tryCount += 1;
     LoggerProxy.logger.log(`ReconnectionManager:index#media --> Begin reestablishment of media at ${new Date()}`);
+    let reCreatePeerConnection = false;
```
Comment (mwegman) [inline packages/node_modules/@webex/plugin-meetings/src/reconnection-manager/index.js:197]: Minor note. Recreate is one word so the variable would be `recreatePeerConnection`

File: packages/node_modules/@webex/plugin-meetings/src/reconnection-manager/index.js [hunk 2]
```diff
@@ -108,7 +108,20 @@ export default class ReconnectionManager {
   reconnect(meeting) {
     const validatedReconnect = this.validate(meeting);
 
-    return validatedReconnect === true ? this.execute(meeting) : Promise.reject(validatedReconnect);
+    if (validatedReconnect === true) {
+      // TODO : find a better place to add this if any
+      if (meeting.locusInfo.parsedLocus.self.joinedWith.state === 'LEFT' && meeting.locusInfo.parsedLocus.self.joinedWith.reason === 'INACTIVE') {
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/reconnection-manager/index.js:113]: `FULL_STATE.LEFT` and `FULL_STATE.INACTIVE` constants are available

File: packages/node_modules/@webex/plugin-meetings/src/reconnection-manager/index.js [hunk 3]
```diff
@@ -180,11 +194,16 @@ export default class ReconnectionManager {
   media(meeting) {
     this.tryCount += 1;
     LoggerProxy.logger.log(`ReconnectionManager:index#media --> Begin reestablishment of media at ${new Date()}`);
+    let reCreatePeerConnection = false;
 
+    if (meeting.mediaProperties.peerConnection.iceConnectionState === 'failed' ||
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/reconnection-manager/index.js:199]: Please use the `ICE_STATE.CLOSED` and `ICE_STATE.FAILED` constants

File: packages/node_modules/@webex/plugin-meetings/src/roap/index.js [hunk 4]
```diff
@@ -220,7 +220,7 @@ export default class Roap extends StatelessWebexPlugin {
    * @memberof Roap
    */
   sendRoapMediaRequest(options) {
-    const {meeting} = options;
+    const {meeting, reconnect} = options;
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/roap/index.js:223]: Please update the `RoapOptions` typedef on line 12 with this reconnect property.

File: packages/node_modules/@webex/plugin-meetings/src/locus-info/selfUtils.js [hunk 5]
```diff
@@ -107,6 +104,15 @@ SelfUtils.getStatus = (status) => ({
   slides: status.videoSlidesStatus
 });
 
+
+/**
+ * @param {Object} oldSelf
+ * @param {Object} changedSelf
+ * @returns {Boolean}
+ */
+SelfUtils.wasMediaInactiveOrReleased = (oldSelf, changedSelf) => oldSelf && oldSelf.joinedWith.state === 'JOINED' && changedSelf.joinedWith.state === 'LEFT' && changedSelf.joinedWith.reason === 'INACTIVE';
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/locus-info/selfUtils.js:113]: Please use constants here.

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 6]
```diff
@@ -597,6 +597,22 @@ export default class Meeting extends StatelessWebexPlugin {
...
+
+  /**
+   * Set up the locus info listener for meetings disconnected due to inactivity
+   * @returns {undefined}
+   * @private
+   * @memberof Meeting
+   */
+  mediaInactiveHandler() {
+    this.locusInfo.on(EVENTS.DISCONNECT_DUE_TO_INACTIVITY, () => {
+      // Thought of passing the self value to determine the left state
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:612]: Add a log statement here.

Comment (adamweeks) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:612]: You can remove these comments, they aren't needed.

**Comment by adamweeks** [comment]
Please use `feat(meetings)` instead of `fix` in your commit message as we're not fixing bugs but adding features.

File: packages/node_modules/@webex/plugin-meetings/src/locus-info/selfUtils.js [hunk 7]
```diff
@@ -107,10 +106,20 @@ SelfUtils.getStatus = (status) => ({
...
+ * @param {Object} changedSelf
  * @returns {Boolean}
  */
+SelfUtils.wasMediaInactiveOrReleased = (oldSelf = {}, changedSelf) => oldSelf.joinedWith && oldSelf.joinedWith.state === _JOINED_ && changedSelf.joinedWith.state === _LEFT_ &&
+ (changedSelf.joinedWith.reason === MEETING_END_REASON.INACTIVE || changedSelf.joinedWith.reason === MEETING_END_REASON.MEDIA_RELEASED);
+
+
+/**
+ * @param {Object} check
+ * @returns {Boolean}ed
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/locus-info/selfUtils.js:121]:
```suggestion
 * @returns {Boolean}
```

File: packages/node_modules/@webex/plugin-meetings/src/reconnection-manager/index.js [hunk 8]
```diff
@@ -143,59 +153,29 @@ export default class ReconnectionManager {
...
-
-
-    // TODO: chekc peerconnection status
-    // create an offer
-    return this.reconnectMedia(meeting)
-      .then((media) => Promise.resolve(media))
-      .catch((err) => {
+          meeting, // or can pass meeting ID
+          reconnect: true // Need to check if its a reconnect after rejoin or media inactivity
+        })).catch((err) => {
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/reconnection-manager/index.js:177]: It looks like the `execute` function relies on this failure to do the retry logic. We're only catching it there though, not elevating the error as a promise rejection on `reconnectMedia`.

File: packages/node_modules/@webex/plugin-meetings/src/reconnection-manager/index.js [hunk 9]
```diff
@@ -205,9 +185,10 @@ export default class ReconnectionManager {
    * @private
    * @memberof ReconnectionManager
    */
-  static cleanup(meeting) {
-    // close pcs, unset to null and reuse the streams with new pcs
+  static setupPeerconnection(meeting) {
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/reconnection-manager/index.js:188]: Minor but we use `PeerConnection` everywhere else in this file, so let's use that capitalization.

File: packages/node_modules/@webex/plugin-meetings/src/constants.js [hunk 10]
```diff
@@ -687,6 +689,12 @@ export const MEETING_ERRORS = {
   NO_HTTP_RESPONSE: 'NO_HTTP_RESPONSE'
 };
 
+
+export const MEETING_END_REASON = {
```
Comment (chrisadubois) [inline packages/node_modules/@webex/plugin-meetings/src/constants.js:693]: 👍

File: packages/node_modules/@webex/plugin-meetings/src/reconnection-manager/index.js [hunk 11]
```diff
@@ -108,7 +108,17 @@ export default class ReconnectionManager {
   reconnect(meeting) {
     const validatedReconnect = this.validate(meeting);
 
-    return validatedReconnect === true ? this.execute(meeting) : Promise.reject(validatedReconnect);
+    if (validatedReconnect === true) {
```
Comment (chrisadubois) [inline packages/node_modules/@webex/plugin-meetings/src/reconnection-manager/index.js:111]: can just change this to validatedReconnect

File: packages/node_modules/@webex/plugin-meetings/src/reconnection-manager/index.js (hunk 8 above)
Comment (chrisadubois) [reply packages/node_modules/@webex/plugin-meetings/src/reconnection-manager/index.js:177]: Agree, we should throw from here

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 12]
```diff
@@ -597,6 +597,21 @@ export default class Meeting extends StatelessWebexPlugin {
...
+
+
+  /**
+   * Set up the locus info listener for meetings disconnected due to inactivity
+   * @returns {undefined}
+   * @private
+   * @memberof Meeting
+   */
+  mediaInactiveHandler() {
+    this.locusInfo.on(EVENTS.DISCONNECT_DUE_TO_INACTIVITY, (res) => {
```
Comment (chrisadubois) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:611]: we should track these type of reconnects as a success and failure in the operational metrics so might be good just to make this location with a TODO

File: packages/node_modules/@webex/plugin-meetings/src/reconnection-manager/index.js (hunk 8 above)
Comment (arun3528) [reply packages/node_modules/@webex/plugin-meetings/src/reconnection-manager/index.js:177]: done

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 13]
```diff
@@ -597,6 +597,22 @@ export default class Meeting extends StatelessWebexPlugin {
     this.setUpLocusMediaSharesListener();
     this.setUpLocusInfoMeetingInfoListener();
     this.setUpLocusInfoAssignHostListener();
+    this.mediaInactiveHandler();
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:600]: Could we name this like the other functions above for consistency? 
```suggestion
    this.setUpLocusInfoMediaInactiveListener();
```

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 14]
```diff
@@ -597,6 +597,22 @@ export default class Meeting extends StatelessWebexPlugin {
...
+  /**
+   * Set up the locus info listener for meetings disconnected due to inactivity
+   * @returns {undefined}
+   * @private
+   * @memberof Meeting
+   */
+  mediaInactiveHandler() {
+    this.locusInfo.on(EVENTS.DISCONNECT_DUE_TO_INACTIVITY, (res) => {
+      // TODO: need to send metric for locus disconnect
+      LoggerProxy.logger.error(`Meeting:index#mediaInactiveHandler --> Meeting disconnected due to ${res.reason}`);
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:613]: Don't we know that this is due to inactivity? What does the reason give us?

```suggestion
      LoggerProxy.logger.error(`Meeting:index#mediaInactiveHandler --> Meeting disconnected due to inactivity: ${res.reason}`);
```

**Comment by adamweeks** [comment]
How do we handle the case where the `meeting.join` fails due to network disconnected?

**Comment by arun3528** [comment]
@adamweeks you are correct if the join fails we cannot retry . So i will move the join function inside the execute so that if its fails it will try again on retry . If its already joined it wont try to join again

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 15]
```diff
@@ -597,6 +597,22 @@ export default class Meeting extends StatelessWebexPlugin {
...
+   * Set up the locus info listener for meetings disconnected due to inactivity
+   * @returns {undefined}
+   * @private
+   * @memberof Meeting
+   */
+  setUpLocusInfoMediaInactiveListener() {
+    this.locusInfo.on(EVENTS.DISCONNECT_DUE_TO_INACTIVITY, (res) => {
+      // TODO: need to send metric for locus disconnect
+      LoggerProxy.logger.error(`Meeting:index#setUpLocusInfoMediaInactiveListener --> Meeting disconnected due to inactivity: ${res.reason}`);
+      this.reconnect();
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:614]: Is it safe to call this function yet? We removed its call from the ice disconnect logic until we had further testing. I'd rather mark this as a TODO for once the reconnect logic is solid.

File: packages/node_modules/@webex/plugin-meetings/src/reconnection-manager/index.js [hunk 16]
```diff
@@ -138,76 +152,48 @@ export default class ReconnectionManager {
...
+      remoteQualityLevel: meeting.mediaProperties.remoteQualityLevel
+    })
+      .then((peerConnection) => meeting.setRemoteStream(peerConnection))
       .then(() => meeting.roap
         .sendRoapMediaRequest({
           sdp: meeting.mediaProperties.peerConnection.sdp,
           roapSeq: meeting.roapSeq,
-          meeting // or can pass meeting ID
+          meeting, // or can pass meeting ID
+          reconnect: true // Need to check if its a reconnect after rejoin or media inactivity
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/reconnection-manager/index.js:180]: What is the check that is needed here?

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js (hunk 15 above)
Comment (arun3528) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:614]: I am guessing there is a reconnect config for meetings plugin we can check that . If our testing goes good we can ask them to enabled the toggle with out a code changes

File: packages/node_modules/@webex/plugin-meetings/src/reconnection-manager/index.js (hunk 16 above)
Comment (arun3528) [reply packages/node_modules/@webex/plugin-meetings/src/reconnection-manager/index.js:254]: this was just a comment

**Pull request merged**